File: 0002-add-two-numbers/main.rb
# Definition for singly-linked list.
class ListNode
    attr_accessor :val, :next
    def initialize(val = 0, _next = nil)
        @val = val
        @next = _next
    end
end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}

def add_two_numbers(l1, l2)
  l1_vals = []
  l2_vals = []

  node = l1
  loop do
    l1_vals.unshift(node.val)
    node = node.next
    break if node.nil?
  end

  node = l2
  loop do
    l2_vals.unshift(node.val)
    node = node.next
    break if node.nil?
  end

  l1_int = l1_vals.join.to_i
  l2_int = l2_vals.join.to_i
  sum = l1_int + l2_int

  digits = sum.to_s.chars.reverse
  n = nil
  digits.each do |c|
    if n.nil?
      n = ListNode.new(c.to_i)
    else
      tmp = n
      until n.next.nil?
        n = n.next
      end
      n.next = ListNode.new(c.to_i)
      n = tmp
    end
  end
  return n
end

l1 = ListNode.new(2, ListNode.new(4, ListNode.new(3)))
l2 = ListNode.new(5, ListNode.new(6, ListNode.new(4)))

r = add_two_numbers(l1, l2)

File: 0002-add-two-numbers/main.js
function ListNode(val, next) {
  this.val = (val===undefined ? 0 : val)
  this.next = (next===undefined ? null : next)
}

var addTwoNumbers = function(l1, l2) {
  let l1_vals = [];
  let l2_vals = [];

  let node = l1;
  while (node) {
    l1_vals.unshift(node.val);
    node = node.next;
  }

  node = l2;
  while (node) {
    l2_vals.unshift(node.val);
    node = node.next;
  }

  l1_int = BigInt(l1_vals.join(''));
  l2_int = BigInt(l2_vals.join(''));

  // console.log(l1_int);
  // console.log(l2_int);

  let sum = l1_int + l2_int;

  // console.log(sum);

  let digits = sum.toString().split('').reverse();
  // console.log(digits);
  let n = null;

  for (let i = 0; i < digits.length; i++) {
    let c = digits[i];
    if (n === null) {
      n = new ListNode(Number(c));
    } else {
      let tmp = n;
      while (n.next) {
        n = n.next;
      }
      n.next = new ListNode(Number(c));
      n = tmp;
    }
  }
  return n;
};

// let l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
// let l2 = new ListNode(5, new ListNode(6, new ListNode(4)));
// let r1 = addTwoNumbers(l1, l2);

let l1 = new ListNode(1, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(0, new ListNode(1)))))))))))))))))))))))))))))))
let l2 = new ListNode(5, new ListNode(6, new ListNode(4)));
let r2 = addTwoNumbers(l1, l2);

console.log(r2);
